internal/app/backend/backend_service.go

```diff
@@ -75,198 +70,194 @@ func (s *backendService) FetchMatches(req *pb.FetchMatchesRequest, stream pb.Bac
+			"mmfErr":             mmfErr,
+		}).Error("error(s) in FetchMatches call.")
+
+		return fmt.Errorf(
+			"Error(s) in FetchMatches call. synchronizeSendErr=[%s], synchronizeRecvErr=[%s], mmfErr=[%s]",
+			synchronizeSendErr,
+			synchronizeRecvErr,
+			mmfErr,
+		)
+	}
+
+	return nil
+}
+
+func synchronizeSend(ctx context.Context, proposals <-chan *pb.Match, syncStream synchronizerStream) error {
+sendProposals:
+	for {
+		select {
+		case <-ctx.Done():
+			break sendProposals
+		case p, ok := <-proposals:
+			if !ok {
 				break sendProposals
-			default:
 			}
-
-			telemetry.RecordUnitMeasurement(stream.Context(), mMatchesSentToEvaluation)
-			err = syncStream.Send(&ipb.SynchronizeRequest{Proposal: p})
+			telemetry.RecordUnitMeasurement(ctx, mMatchesSentToEvaluation)
+			err := syncStream.Send(&ipb.SynchronizeRequest{Proposal: p})
 			if err != nil {
 				return fmt.Errorf("error sending proposal to synchronizer: %w", err)
 			}
 		}
-
-		err = syncStream.CloseSend()
-		if err != nil {
-			return fmt.Errorf("error closing send stream of proposals to synchronizer: %w", err)
-		}
-		return nil
 	}
 
-	go func() {
-		var startMmfsOnce sync.Once
-		defer func() {
-			startMmfsOnce.Do(func() {
-				errors <- fmt.Errorf("MMFS were never started")
-			})
-		}()
+	err := syncStream.CloseSend()
+	if err != nil {
+		return fmt.Errorf("error closing send stream of proposals to synchronizer: %w", err)
+	}
+	return nil
+}
 
-		for {
-			resp, err := syncStream.Recv()
-			if err == io.EOF {
-				errors <- nil
-				return
-			}
-			if err != nil {
-				errors <- fmt.Errorf("error receiving match from synchronizer: %w", err)
-				return
-			}
+func synchronizeRecv(syncStream synchronizerStream, stream pb.BackendService_FetchMatchesServer, startMmfs chan<- struct{}, cancelMmfs context.CancelFunc) error {
+	var startMmfsOnce sync.Once
 
-			if resp.StartMmfs {
-				go startMmfsOnce.Do(func() {
-					errors <- startMmfs()
-				})
-			}
+	for {
+		resp, err := syncStream.Recv()
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("error receiving match from synchronizer: %w", err)
+		}
 
-			if resp.CancelMmfs {
-				cancelMmfs()
-			}
+		if resp.StartMmfs {
+			go startMmfsOnce.Do(func() {
+				close(startMmfs)
+			})
+		}
 
-			if resp.Match != nil {
-				telemetry.RecordUnitMeasurement(stream.Context(), mMatchesFetched)
-				err = stream.Send(&pb.FetchMatchesResponse{Match: resp.Match})
-				if err != nil {
-					errors <- fmt.Errorf("error sending match to caller of backend: %w", err)
-					return
-				}
-			}
+		if resp.CancelMmfs {
+			cancelMmfs()
 		}
-	}()
 
-	for i := 0; i < 2; i++ {
-		err := <-errors
-		if err != nil {
-			logger.WithFields(logrus.Fields{
-				"error": err.Error(),
-			}).Error("error in FetchMatches call.")
-			return err
+		if resp.Match != nil {
+			telemetry.RecordUnitMeasurement(stream.Context(), mMatchesFetched)
+			err = stream.Send(&pb.FetchMatchesResponse{Match: resp.Match})
```

Comment:
nit: resp.GetMatch()

Reply:
err is nil, so resp is not nil.  There's also the above line resp.Match != nil, which guarantees there to be a value here.

Reply:
I understand the code logic, but I am more concerned about using the proto getters as a standard in our repo.